cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/NamespaceSourceControlMetadataRefreshService.java

```diff
@@ -0,0 +1,314 @@
+    if (executor != null) {
+      executor.shutdownNow();
+    }
+  }
+
+  private void updateSourceControlMeta(ApplicationReference appRef, String repoFileHash,
+      Long refreshStartTime) {
+    TransactionRunners.run(transactionRunner, context -> {
+      RepositorySourceControlMetadataStore repoMetadataStore = getRepoSourceControlMetadataStore(
+          context);
+      NamespaceSourceControlMetadataStore namespaceMetadataStore =
+          getNamespaceSourceControlMetadataStore(context);
+      SourceControlMeta namespaceSourceControlMeta = namespaceMetadataStore.get(appRef);
+
+      if (namespaceSourceControlMeta == null) {
+        repoMetadataStore.write(appRef, false, 0L);
+        return;
+      }
+
+      if (namespaceSourceControlMeta.getLastSyncedAt() != null
+          && namespaceSourceControlMeta.getLastSyncedAt().isAfter(
+          Instant.ofEpochMilli(refreshStartTime))) {
+        repoMetadataStore.write(appRef,
+            Boolean.TRUE.equals(namespaceSourceControlMeta.getSyncStatus()),
+            namespaceSourceControlMeta.getLastSyncedAt().toEpochMilli());
+        return;
+      }
+      boolean isSynced = repoFileHash.equals(namespaceSourceControlMeta.getFileHash());
+      namespaceMetadataStore.write(appRef,
+          SourceControlMeta.builder(namespaceSourceControlMeta).setSyncStatus(isSynced).build());
+      repoMetadataStore.write(appRef,
+          isSynced, namespaceSourceControlMeta.getLastSyncedAt() == null ? 0L :
+              namespaceSourceControlMeta.getLastSyncedAt().toEpochMilli());
+    });
+  }
+
+  private void cleanupRepoSourceControlMetaInBatches(String namespace, HashSet<String> repoFiles) {
+    TransactionRunners.run(transactionRunner, context -> {
+      RepositorySourceControlMetadataStore store = getRepoSourceControlMetadataStore(context);
+      ScanSourceControlMetadataRequest request = ScanSourceControlMetadataRequest
+          .builder()
+          .setNamespace(namespace)
+          .setLimit(Integer.MAX_VALUE)
+          .build();
+      ArrayList<SourceControlMetadataRecord> records = new ArrayList<>();
+      String type = EntityType.APPLICATION.toString();
+
+      store.scan(request, type, record -> {
+        records.add(record);
+        if (records.size() == batchSize) {
+          try {
+            cleanUpRecords(store, repoFiles, records);
+          } catch (IOException e) {
+            throw new RuntimeException(e);
+          }
+          records.clear();
+        }
+      });
+
+      if (!records.isEmpty()) {
+        cleanUpRecords(store, repoFiles, records);
+      }
+    });
+  }
+
+
+  private void cleanUpRecords(RepositorySourceControlMetadataStore store, HashSet<String> repoFiles,
+      List<SourceControlMetadataRecord> records)
+      throws IOException {
+    for (SourceControlMetadataRecord record : records) {
+      if (!repoFiles.contains(record.getName())) {
+        store.delete(new ApplicationReference(record.getNamespace(), record.getName()));
```

Comment:
What does the `store.delete` actually do? Asking it because this method is nested inside the transactional scan method.